original/man/man3/beliefnetdata.3
.TH "beliefnetdata" 3 "28 Jul 2003" "VFML" \" -*- nroff -*-
.ad l
.nh
.SH NAME
beliefnetdata \- 
.SH "Detailed Description"
.PP 
Creates a data set by sampling from a Bayesian Network. 

>This program creates a synthetic data set by loading an existing belief network (in \fCBIF format\fP) and sampling from it, possibly introducing noise. This program will also create a .names file for the resulting data set. More specifically, belifnetdata reads the network, generates samples from it with \fBBNGenerateSample\fP, and adds noise to them (if requested) with \fBExampleAddNoise\fP.
.PP
Multiple runs with the same seed parameter produce the same results. Also note that running this command with one level of -v will output some statistics about the belif net which you might find useful.
.PP
VFML comes with a collection of benchmark belief nets, and you may want \fCmore information on these\fP.
.PP
.SS "Arguments"
.PP
.IP "\(bu" 2
-f 'stem name for output'
.IP "  \(bu" 4
(default DF)
.PP

.IP "\(bu" 2
-bnf 'file containing belief net'
.IP "  \(bu" 4
(default DF.bif)
.PP

.IP "\(bu" 2
-train 'size of training set'
.IP "  \(bu" 4
(default 10000)
.PP

.IP "\(bu" 2
-infinite
.IP "  \(bu" 4
Generate an infinite stream of training examples, overrides -train flags, only makes sense with -stdout (default off)
.PP

.IP "\(bu" 2
-stdout
.IP "  \(bu" 4
Output the trainset to stdout (default to 'stem>'data)
.PP

.IP "\(bu" 2
-seed 'random seed'
.IP "  \(bu" 4
(default to random)
.PP

.IP "\(bu" 2
-noise 'noise level'
.IP "  \(bu" 4
10 means change 10% of values, including classes (default to 0)
.PP

.IP "\(bu" 2
-target 'dir'
.IP "  \(bu" 4
Set the output directory (default '.')
.PP

.IP "\(bu" 2
-v
.IP "  \(bu" 4
Increase the message level
.PP

.IP "\(bu" 2
-h
.IP "  \(bu" 4
Run with this argument to get a list of arguments and their meanings.
.br

.PP

.PP
.PP
.SS "Example"
.PP
\fC\fP
.PP
\fCbeliefnetdata -f train -bnf alarm.bif -train 1000 -seed 111 -noise 5\fP
.PP
Creates 1000 samples from the alarm network, randomly corrupts 5% of their values, write the resulting samples to train.data (and create a file train.names) and reproduce the same data set everytime the same arguments are used (thanks to the seed parameter)
.PP
.SH SYNOPSIS
.br
.PP
.SH "Author"
.PP 
Generated automatically by Doxygen for VFML from the source code.


original/src/tools/beliefnetdata.c
#include "vfml.h"
#include <stdio.h>
#include <string.h>
#include <sys/times.h>
#include <time.h>
#include <math.h>

char  *gFileStem         = "DF";
char  *gBeliefNetFile    = "DF.bif";
int   gInfiniteStream    = 0;
char  *gTargetDirectory  = ".";
unsigned long  gNumTrainExamples = 10000;
float gErrorLevel = 0.00;


long  gSeed = -1;
int   gStdout = 0;

//static char *_AllocateString(char *first) {
//   char *second = MNewPtr(sizeof(char) * (strlen(first) + 1));
//
//   strcpy(second, first);
//
//   return second;
//}

static void _processArgs(int argc, char *argv[]) {
   int i;

   /* HERE on the ones that use the next arg make sure it is there */
   for(i = 1 ; i < argc ; i++) {
      if(!strcmp(argv[i], "-f")) {
         gFileStem = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-bnf")) {
         gBeliefNetFile = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-h")) {
         printf("Generate a synthetic dataset\n");
         printf("   -f <stem name for output> (default DF)\n");
         printf("   -bnf <file containing belief net> (default DF.bif)\n");
         printf("   -train <size of training set> (default 10000)\n");
         printf("   -infinite\t Generate an infinite stream of training examples,\n\t overrides -train flags, only makes sense with -stdout (default off)\n");
         printf("   -stdout  output the trainset to stdout (default to <stem>.data)\n");
         printf("   -seed <random seed> (default to random)\n");
         printf("   -noise <noise level> 10 means change 10%% of values, including \n\t classes (default to 0)\n");
         printf("   -target <dir> Set the output directory (default '.')\n");
         printf("   -v increase the message level\n");
         exit(0);
      } else if(!strcmp(argv[i], "-train")) {
         sscanf(argv[i+1], "%lu", &gNumTrainExamples);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-seed")) {
         sscanf(argv[i+1], "%lu", &gSeed);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-noise")) {
         sscanf(argv[i+1], "%f", &gErrorLevel);
         gErrorLevel /= 100.0;
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-v")) {
         DebugSetMessageLevel(DebugGetMessageLevel() + 1);
      } else if(!strcmp(argv[i], "-stdout")) {
         gStdout = 1;
      } else if(!strcmp(argv[i], "-infinite")) {
         gInfiniteStream = 1;
      } else if(!strcmp(argv[i], "-target")) {
         gTargetDirectory = argv[i+1];
         /* ignore the next argument */
         i++;
      } else {
         printf("unknown argument '%s', use -h for help\n", argv[i]);
         exit(0);
      }
   }
}


static BeliefNet _readBN(void) {
   char fileName[255];

   sprintf(fileName, "%s", gBeliefNetFile);

   return BNReadBIF(fileName);
}

static void _writeSpec(BeliefNet bn) {
   char fileName[255];
   FILE *fileOut;

   sprintf(fileName, "%s/%s.names", gTargetDirectory, gFileStem);
   fileOut = fopen(fileName, "w");
   ExampleSpecWrite(BNGetExampleSpec(bn), fileOut);
   //BNWriteBIF(bn, fileOut);

   /* dunno if this flush helps, but sometimes the learner tries
       to open this before the file appears when piping the output
       to the learner via stdout.  Maybe delay a second after this 
       close? */

   fflush(fileOut);
   fclose(fileOut);
}

static void _makeData(BeliefNet bn) {
   char fileNames[255];
   FILE *fileOut;
   ExamplePtr e;
   long i;

   if(!gStdout) {
      sprintf(fileNames, "%s/%s.data", gTargetDirectory, gFileStem);
      fileOut = fopen(fileNames, "w");
   } else {
      fileOut = stdout;
   }

   for(i = 0 ; i < gNumTrainExamples || gInfiniteStream ; i++) {
      e = BNGenerateSample(bn);
      ExampleAddNoise(e, gErrorLevel, 1, 1);
      ExampleWrite(e, fileOut);
      ExampleFree(e);
   }

   if(!gStdout) {
      fclose(fileOut);
   }
}

int main(int argc, char *argv[]) {
   BeliefNet bn;
   _processArgs(argc, argv);

   /* randomize the random # generator */
   RandomInit();
   /* seed for the data */
   if(gSeed != -1) {
      RandomSeed(gSeed);
   } else {
      gSeed = RandomRange(1, 30000);
      RandomSeed(gSeed);
   }

   bn = _readBN();

   DebugError(bn == 0, "Unable to read .bif file.\n");

   if(DebugGetMessageLevel() > 0) {
      BNPrintStats(bn);
   }

   _writeSpec(bn);
   _makeData(bn);

   return 0;
}
